local_gemma/attention.py

```diff
@@ -0,0 +1,143 @@
+        self.num_key_value_groups = self.num_heads // self.num_key_value_heads
+        self.max_position_embeddings = config.max_position_embeddings
+        self.rope_theta = config.rope_theta
+        self.is_causal = True
+        self.scaling = config.query_pre_attn_scalar**-0.5
+
+        if self.hidden_size % self.num_heads != 0:
+            raise ValueError(
+                f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
+                f" and `num_heads`: {self.num_heads})."
+            )
+
+        # fused attention proj
+        total_head_dim = (2 * self.num_key_value_heads + self.num_heads) * self.head_dim
+        self.qkv_proj = nn.Linear(self.hidden_size, total_head_dim, bias=config.attention_bias)
+
+        # conversion from un-fused to fused
+        self._register_load_state_dict_pre_hook(self.load_hook)
+
+        self.o_proj = nn.Linear(self.num_heads * self.head_dim, self.hidden_size, bias=config.attention_bias)
+        self.rotary_emb = Gemma2RotaryEmbedding(
+            self.head_dim,
+            max_position_embeddings=self.max_position_embeddings,
+            base=self.rope_theta,
+        )
+        self.sliding_window = config.sliding_window if not bool(layer_idx % 2) else None
+
+    def load_hook(self, state_dict, prefix, *args):
+        if prefix + "q_proj.weight" in state_dict:
+            q_proj = state_dict.pop(prefix + "q_proj.weight")
+            k_proj = state_dict.pop(prefix + "k_proj.weight")
+            v_proj = state_dict.pop(prefix + "v_proj.weight")
+            state_dict[prefix + "qkv_proj.weight"] = torch.cat([q_proj, k_proj, v_proj])
+
+            if self.config.attention_bias:
+                q_bias = state_dict.pop(prefix + "q_proj.bias")
+                k_bias = state_dict.pop(prefix + "k_proj.bias")
+                v_bias = state_dict.pop(prefix + "v_proj.bias")
+                state_dict[prefix + "qkv_proj.bias"] = torch.cat([q_bias, k_bias, v_bias])
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Cache] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+        cache_position: Optional[torch.LongTensor] = None,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        bsz, q_len, _ = hidden_states.size()
+
+        query_key_value_states = self.qkv_proj(hidden_states)
+        query_size = self.num_heads * self.head_dim
+        key_value_size = self.num_key_value_heads * self.head_dim
+        query_states, key_states, value_states = query_key_value_states.split(
+            [query_size, key_value_size, key_value_size], dim=-1
+        )
+
+        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
```

Comment:
The remainder of this attention class is one-to-one the same as that in Transformers: https://github.com/huggingface/transformers/blob/535fe78b9f1d148684723e51f00645351880c47a/src/transformers/models/gemma2/modeling_gemma2.py#L224

We'd ideally have a mechanism to keep the attention modules sync'd as a follow-up PR